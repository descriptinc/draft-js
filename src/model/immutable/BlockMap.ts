/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {BlockNodeRecord} from './BlockNodeRecord';
import {flatMap, flatten} from '../descript/Iterables';

export type BlockMap = ReadonlyMap<string, BlockNodeRecord>;

// FIXME [mvp]: dealing with order of `updates`

export function mergeMapUpdates<T>(
  originalMap: ReadonlyMap<string, T>,
  updates: Record<string, T | null>,
): ReadonlyMap<string, T> {
  let didChange = false;
  const seenKeys = new Set<string>();
  const result: ReadonlyMap<string, T> = new Map(
    flatten([
      flatMap(originalMap, (entry): [string, T] | undefined => {
        const [key, existingBlock] = entry;
        seenKeys.add(key);
        const newBlock = updates[key];
        if (newBlock === null) {
          // remove this block from the map
          return undefined;
        }
        if (!newBlock || newBlock === existingBlock) {
          return entry;
        }
        didChange = true;
        return [key, newBlock];
      }),
      flatMap(Object.entries(updates), (entry): [string, T] | undefined => {
        const [key, block] = entry;
        if (!block || seenKeys.has(key)) {
          return undefined;
        }
        didChange = true;
        return [key, block];
      }),
    ]),
  );
  return didChange ? result : originalMap;
}
