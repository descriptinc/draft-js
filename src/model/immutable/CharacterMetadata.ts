/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftInlineStyle} from './DraftInlineStyle';
import {getKey2} from '../descript/cache';

export type CharacterMetadata = Readonly<{
  entity: string | null;
  style: DraftInlineStyle;
}>;

export function applyStyle(
  meta: CharacterMetadata,
  style: string,
): CharacterMetadata {
  if (meta.style.has(style)) {
    return meta;
  }
  return {
    ...meta,
    style: new Set([...meta.style.keys(), style]),
  };
}

export function removeStyle(
  meta: CharacterMetadata,
  style: string,
): CharacterMetadata {
  if (!meta.style.has(style)) {
    return meta;
  }
  const without = new Set(meta.style);
  without.delete(style);
  return {
    ...meta,
    style: without,
  };
}

export function applyEntity(
  meta: CharacterMetadata,
  entity: string | null,
): CharacterMetadata {
  if (meta.entity === entity) {
    return meta;
  }
  return {
    ...meta,
    entity,
  };
}

const pool: Record<string, CharacterMetadata> = {};
export const EMPTY_SET: DraftInlineStyle = new Set();
export const EMPTY_CHARACTER: CharacterMetadata = {
  entity: null,
  style: EMPTY_SET,
};

export function makeCharacterMetadata(
  config?: Partial<CharacterMetadata>,
): CharacterMetadata {
  if (!config) {
    return EMPTY_CHARACTER;
  }
  const entity = config.entity || null;
  const style = config.style || new Set();
  const styles = [...style];
  styles.sort(); // FIXME [perf]: is this sort too expensive?
  const styleStr = styles.join(',');
  const key = getKey2(entity || '', styleStr);
  const existing = pool[key];
  if (existing) {
    return existing;
  }
  const char = {entity, style};
  pool[key] = char;
  return char;
}
