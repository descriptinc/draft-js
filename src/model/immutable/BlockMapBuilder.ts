/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {BlockNodeRecord} from './BlockNodeRecord';
import {BlockMap} from './BlockMap';
import {map} from '../descript/Iterables';

export function createFromArray(blocks: Iterable<BlockNodeRecord>): BlockMap {
  return new Map(
    map(blocks, (block): [string, BlockNodeRecord] => [block.key, block]),
  );
}
