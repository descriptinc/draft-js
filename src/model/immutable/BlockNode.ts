/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftBlockType} from '../constants/DraftBlockType';
import {CharacterMetadata} from './CharacterMetadata';

export type BlockNodeKey = string;

export type BlockNode = Readonly<{
  characterList: readonly CharacterMetadata[];
  data: Readonly<Record<string, any>>;
  depth: number;
  key: BlockNodeKey;
  text: string;
  type: DraftBlockType;
}>;
