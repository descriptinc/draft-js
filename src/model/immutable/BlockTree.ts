/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */

'use strict';

import {ContentState} from './ContentState';
import {BlockNodeRecord} from './BlockNodeRecord';
import {DraftDecoratorType} from '../decorators/DraftDecoratorType';
import {getRange} from '../descript/RangeCache';
import {findRangesImmutable} from './findRangesImmutable';
import {map} from '../descript/Iterables';
import {CharacterMetadata} from './CharacterMetadata';

const returnTrue = function() {
  return true;
};

type LeafRange = Readonly<{
  start: number;
  end: number;
}>;

type DecoratorRange = LeafRange &
  Readonly<{
    decoratorKey: string | null;
    leaves: readonly LeafRange[];
  }>;

function makeDecoratorRange(
  start: number,
  end: number,
  decoratorKey: string | null,
  leaves: readonly LeafRange[],
): DecoratorRange {
  return {
    start,
    end,
    decoratorKey,
    leaves,
  };
}

const BlockTree = {
  /**
   * Generate a block tree for a given ContentBlock/decorator pair.
   */
  generate: function(
    contentState: ContentState,
    block: BlockNodeRecord,
    decorator: DraftDecoratorType | null,
  ): readonly DecoratorRange[] {
    const textLength = block.text.length;
    if (!textLength) {
      return [makeDecoratorRange(0, 0, null, [getRange(0, 0)])];
    }

    const leafSets: DecoratorRange[] = [];
    const decorations = decorator
      ? decorator.getDecorations(block, contentState)
      : new Array(textLength).fill(null);

    const chars = block.characterList;

    findRangesImmutable(decorations, areEqual, returnTrue, (start, end) => {
      leafSets.push(
        makeDecoratorRange(
          start,
          end,
          decorations[start],
          generateLeaves(chars.slice(start, end), start),
        ),
      );
    });

    return leafSets;
  },
};

/**
 * Generate LeafRange records for a given character list.
 */
function generateLeaves(
  characters: readonly CharacterMetadata[],
  offset: number,
): readonly LeafRange[] {
  const leaves: LeafRange[] = [];
  findRangesImmutable(
    map(characters, ({style}) => style),
    areEqual,
    returnTrue,
    (start, end) => {
      leaves.push(getRange(start + offset, end + offset));
    },
  );
  return leaves;
}

function areEqual(a: any, b: any): boolean {
  return a === b;
}

module.exports = BlockTree;
