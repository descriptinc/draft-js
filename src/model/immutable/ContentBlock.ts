/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {
  CharacterMetadata,
  EMPTY_CHARACTER,
  EMPTY_SET,
} from './CharacterMetadata';
import {BlockNode} from './BlockNode';
import {repeat} from '../descript/Iterables';
import {findRangesImmutable} from './findRangesImmutable';
import {DraftInlineStyle} from './DraftInlineStyle';
import fastDeepEqual from 'fast-deep-equal/es6';

const EMPTY_DATA = {};

export type ContentBlock = BlockNode;

export function makeContentBlock({
  data = EMPTY_DATA,
  key,
  text = '',
  depth = 0,
  type = 'unstyled',
  characterList = Array.from(repeat(text.length, EMPTY_CHARACTER)),
}: Partial<ContentBlock> & {key: string}): ContentBlock {
  return {
    characterList,
    data,
    key,
    text,
    depth,
    type,
  };
}

const haveEqualStyle = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => fastDeepEqual(charA.style, charB.style);

const haveEqualEntity = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => charA.entity === charB.entity;

export function findStyleRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualStyle, filterFn, callback);
}

export function findEntityRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualEntity, filterFn, callback);
}

export function getInlineStyleAt(
  block: BlockNode,
  index: number,
): DraftInlineStyle {
  return block.characterList[index]?.style || EMPTY_SET;
}

export function getEntityAt(block: BlockNode, index: number): string | null {
  return block.characterList[index]?.entity || null;
}
