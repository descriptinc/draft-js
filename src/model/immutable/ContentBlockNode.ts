/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 *
 * This file is a fork of ContentBlock adding support for nesting references by
 * providing links to children, parent, prevSibling, and nextSibling.
 *
 * This is unstable and not part of the public API and should not be used by
 * production systems. This file may be update/removed without notice.
 */

import {BlockNode, BlockNodeKey} from './BlockNode';
import {
  CharacterMetadata,
  EMPTY_CHARACTER,
  EMPTY_SET,
} from './CharacterMetadata';
import fastDeepEqual from 'fast-deep-equal/es6';
import {findRangesImmutable} from './findRangesImmutable';
import {DraftInlineStyle} from './DraftInlineStyle';

const EMPTY_ARRAY: readonly BlockNodeKey[] = [];
const EMPTY_MAP = new Map();

export type ContentBlockNode = BlockNode &
  Readonly<{
    children: readonly BlockNodeKey[];
    parent: BlockNodeKey | null;
    prevSibling: BlockNodeKey | null;
    nextSibling: BlockNodeKey | null;
  }>;

const haveEqualStyle = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => fastDeepEqual(charA.style, charB.style);

const haveEqualEntity = (
  charA: CharacterMetadata,
  charB: CharacterMetadata,
): boolean => charA.entity === charB.entity;

export function makeContentBlockNode({
  parent = null,
  data = EMPTY_MAP,
  key = '',
  text = '',
  depth = 0,
  type = 'unstyled',
  children = EMPTY_ARRAY,
  prevSibling = null,
  nextSibling = null,
  characterList = new Array(text.length).fill(EMPTY_CHARACTER),
}: Partial<ContentBlockNode>): ContentBlockNode {
  return {
    parent,
    characterList,
    data,
    key,
    text,
    depth,
    type,
    children,
    prevSibling,
    nextSibling,
  };
}

export function findStyleRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualStyle, filterFn, callback);
}

export function findEntityRanges(
  block: BlockNode,
  filterFn: (value: CharacterMetadata) => boolean,
  callback: (start: number, end: number) => void,
): void {
  findRangesImmutable(block.characterList, haveEqualEntity, filterFn, callback);
}

export function getInlineStyleAt(
  block: BlockNode,
  index: number,
): DraftInlineStyle {
  return block.characterList[index]?.style || EMPTY_SET;
}

export function getEntityAt(block: BlockNode, index: number): string | null {
  return block.characterList[index]?.entity || null;
}
