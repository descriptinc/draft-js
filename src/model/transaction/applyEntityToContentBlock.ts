/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {applyEntity, CharacterMetadata} from '../immutable/CharacterMetadata';
import {BlockNode} from '../immutable/BlockNode';

export default function applyEntityToContentBlock(
  contentBlock: BlockNode,
  startArg: number,
  end: number,
  entityKey: string | null,
): BlockNode {
  let start = startArg;
  if (start >= end) {
    return contentBlock;
  }

  const existingCharacterList = contentBlock.characterList;
  let characterList: CharacterMetadata[] | undefined;

  while (start < end) {
    if (!characterList && existingCharacterList[start].entity !== entityKey) {
      characterList = Array.from(existingCharacterList);
    }
    if (characterList) {
      characterList[start] = applyEntity(characterList[start], entityKey);
    }
    start++;
  }
  return characterList
    ? {
        ...contentBlock,
        characterList,
      }
    : contentBlock;
}

export function applyEntityToMutableCharacterList(
  characterList: CharacterMetadata[],
  start: number,
  end: number,
  entityKey: string | null,
): void {
  for (let i = start; i < end; i++) {
    if (characterList[i].entity !== entityKey) {
      characterList[i] = applyEntity(characterList[i], entityKey);
    }
  }
}
