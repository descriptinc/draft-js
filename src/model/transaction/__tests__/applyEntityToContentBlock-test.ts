/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 * @format
 */

import {makeContentBlock} from '../../immutable/ContentBlock';
import applyEntityToContentBlock, {
  applyEntityToMutableCharacterList,
} from '../applyEntityToContentBlock';
import {blockToJson} from '../../../util/blockMapToJson';

const sampleBlock = makeContentBlock({
  key: 'a',
  text: 'Hello',
});

const assertApplyEntityToContentBlock = (
  start: number,
  end: number,
  entityKey = 'x',
  contentBlock = sampleBlock,
) => {
  expect(
    blockToJson(applyEntityToContentBlock(contentBlock, start, end, entityKey)),
  ).toMatchSnapshot();
};

test('must apply from the start', () => {
  assertApplyEntityToContentBlock(0, 2);
});

test('must apply within', () => {
  assertApplyEntityToContentBlock(1, 4);
});

test('must apply at the end', () => {
  assertApplyEntityToContentBlock(3, 5);
});

test('must apply to the entire text', () => {
  assertApplyEntityToContentBlock(0, 5);
});

test('must apply to mutable character list', () => {
  const charList = Array.from(sampleBlock.characterList);
  applyEntityToMutableCharacterList(charList, 1, 4, 'x');
  expect(charList).toMatchSnapshot();

  applyEntityToMutableCharacterList(charList, 2, 3, 'y');
  expect(charList).toMatchSnapshot();
});
