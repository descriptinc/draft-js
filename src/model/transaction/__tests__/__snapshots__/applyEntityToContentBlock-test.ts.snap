// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply at the end 1`] = `
Object {
  "characterList": Array [
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
  ],
  "data": Object {},
  "depth": 0,
  "key": "a",
  "text": "Hello",
  "type": "unstyled",
}
`;

exports[`must apply from the start 1`] = `
Object {
  "characterList": Array [
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
  ],
  "data": Object {},
  "depth": 0,
  "key": "a",
  "text": "Hello",
  "type": "unstyled",
}
`;

exports[`must apply to mutable character list 1`] = `
Array [
  Object {
    "entity": null,
    "style": Set {},
  },
  Object {
    "entity": "x",
    "style": Set {},
  },
  Object {
    "entity": "x",
    "style": Set {},
  },
  Object {
    "entity": "x",
    "style": Set {},
  },
  Object {
    "entity": null,
    "style": Set {},
  },
]
`;

exports[`must apply to mutable character list 2`] = `
Array [
  Object {
    "entity": null,
    "style": Set {},
  },
  Object {
    "entity": "x",
    "style": Set {},
  },
  Object {
    "entity": "y",
    "style": Set {},
  },
  Object {
    "entity": "x",
    "style": Set {},
  },
  Object {
    "entity": null,
    "style": Set {},
  },
]
`;

exports[`must apply to the entire text 1`] = `
Object {
  "characterList": Array [
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
  ],
  "data": Object {},
  "depth": 0,
  "key": "a",
  "text": "Hello",
  "type": "unstyled",
}
`;

exports[`must apply within 1`] = `
Object {
  "characterList": Array [
    Object {
      "entity": null,
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": "x",
      "style": Array [],
    },
    Object {
      "entity": null,
      "style": Array [],
    },
  ],
  "data": Object {},
  "depth": 0,
  "key": "a",
  "text": "Hello",
  "type": "unstyled",
}
`;
